### PR TITLE
DEVEX-1668 dxfile api route optimization

### PR DESCRIPTION
**Pull request by mhrvol**
Now using DxFindDataObjects instead of describe

Using project param in a call when all files in bulk describe have only 1 projects, otherwise project is not specified - do we prefer different solution?

I also updated test, as results were coming reversed (probably just different sorting), is this a problem or this does not matter?

Also I created new method in DxFindDataObjects class, which calls the API given in argument, it does not create request same way as apply does. I was not sure how to (easily) use apply method for these purposes, as dxproject is mandatory and I was not sure how use ID as a way to describe a file.

### CODE REVIEW

**Comment by commandlinegirl** [comment]
@orodeh Is there a way for us to benchmark the performance after this update (i.e. have you been doing it)? Thx!

File: src/test/scala/dxWDL/dx/DxdaManifestTest.scala [hunk 1]
```diff
@@ -55,7 +55,7 @@ class DxdaManifestTest extends FlatSpec with Matchers {
             "folder" -> JsString(local.toFile().getParent())
         )
         JsObject(fields)
-    }.toVector
+    }.toVector.reverse
```
Comment (commandlinegirl) [inline src/test/scala/dxWDL/dx/DxdaManifestTest.scala:58]: Why is the sort order reversed? Does `system/findDataObjects` return objects sorted by a different field than `system/describeDataObjects`?

File: src/main/scala/dxWDL/dx/DxFindDataObjects.scala [hunk 2]
```diff
@@ -144,24 +144,33 @@ case class DxFindDataObjects(limit: Option[Int], verbose: Verbose) {
...
       propertyConstraints: Vector[String],
       nameConstraints: Vector[String],
-      withInputOutputSpec: Boolean
-  ): (Map[DxDataObject, DxObjectDescribe], Option[JsValue]) = {
+      withInputOutputSpec: Boolean,
+      idConstraints: Vector[String],
+      extraFields: Vector[String]
+  ): JsObject = {
     var fields = Set(Field.Name, Field.Folder, Field.Size, Field.ArchivalState, Field.Properties)
+    fields ++= extraFields.map(field => Field.withName(field)).toSet // TODO: This is case sensitive, is this ok?
```
Comment (commandlinegirl) [inline src/main/scala/dxWDL/dx/DxFindDataObjects.scala:160]: IMO the user will expect they can supply the field name strings with casing as defined by the `system/findDataObjects` method, so this mapping would then fail. Would it make sense to have a Vector or Set of Field values? Like [here](https://github.com/dnanexus/dxWDL/blob/master/src/main/scala/dxWDL/dx/DxFile.scala#L48) for example

File: src/main/scala/dxWDL/dx/package.scala [hunk 3]
```diff
@@ -117,6 +117,7 @@ object Field extends Enumeration {
       DeveloperNotes, Folder, Id, IgnoreReuse, Inputs, InputSpec, Modified, Name, Outputs,
       OutputSpec, ParentJob, Parts, Project, Properties, RunSpec, Size, Stages, Summary, Tags,
       Title, Types, Version = Value
+  def withNameOpt(s: String): Option[Value] = values.find(_.toString == s)
```
[the remark this answers is not in the document]
Comment (mhrvol) [reply src/main/scala/dxWDL/dx/package.scala:120]: I found enum.withName in the end, at first I did not know it exists, therefore I created this method.

I'll delete it, thanks

File: src/main/scala/dxWDL/dx/DxFindDataObjects.scala [hunk 4]
```diff
@@ -144,24 +144,33 @@ case class DxFindDataObjects(limit: Option[Int], verbose: Verbose) {
     JsObject(part1 ++ part2 ++ part3)
   }
 
-  // Submit a request for a limited number of objects
-  private def submitRequest(
-      scope: JsValue,
-      dxProject: DxProject,
+  // Create a request for a limited number of objects
+  private def createRequest(
```
[the remark this answers is not in the document]
Comment (mhrvol) [reply src/main/scala/dxWDL/dx/DxFindDataObjects.scala:148]: Sure, fixed

File: src/main/scala/dxWDL/dx/DxFindDataObjects.scala [hunk 5]
```diff
@@ -237,23 +256,42 @@ case class DxFindDataObjects(limit: Option[Int], verbose: Verbose) {
             withProperties: Vector[String], // object must have these properties
             nameConstraints: Vector[String], // the object name has to be one of these strings
             withInputOutputSpec: Boolean // should the IO spec be described?
+           ): Map[DxDataObject, DxObjectDescribe] = {
+    apply(Some(dxProject), folder, recurse, klassRestriction, withProperties, nameConstraints, withInputOutputSpec, Vector.empty, extrafields = Vector.empty)
+  }
+
+  def apply(dxProject: Option[DxProject],
```
[the remark this answers is not in the document]
Comment (mhrvol) [reply src/main/scala/dxWDL/dx/DxFindDataObjects.scala:263]: Sure, changed (hopefully) uses of this method in dxWDL.

File: src/main/scala/dxWDL/dx/DxFindDataObjects.scala (hunk 2 above)
Comment (mhrvol) [reply src/main/scala/dxWDL/dx/DxFindDataObjects.scala]: Fixed.

File: src/test/scala/dxWDL/dx/DxdaManifestTest.scala (hunk 1 above)
Comment (mhrvol) [reply src/test/scala/dxWDL/dx/DxdaManifestTest.scala:58]: I am not sure according to which criteria it returns the files, but they seem to be reversed.
I add few more testcases for bulk describe.

Is it ordering of elements in returned map relevant?

File: src/test/scala/dxWDL/dx/DxFileTest.scala [hunk 6]
```diff
@@ -0,0 +1,73 @@
+package dxWDL.dx
```
Comment (commandlinegirl) [inline src/test/scala/dxWDL/dx/DxFileTest.scala:1]: Thanks for adding the unit tests! Would it be possible to add a case where an `extraFields` parameter is passed, for example with "details" or/and "parts" and they are also returned in the describe result.

File: src/test/scala/dxWDL/dx/DxdaManifestTest.scala (hunk 1 above)
Comment (commandlinegirl) [reply src/test/scala/dxWDL/dx/DxdaManifestTest.scala:58]: I don't think it's relevant, no. Was just wondering where it comes from.

**Pull request merged**